### PR TITLE
fix: EthAPI: use StateCompute for feeHistory; apply minimum gas premium

### PR DESCRIPTION
**Pull request by ychiaoli18**
## Related Issues
https://github.com/filecoin-project/lotus/issues/10393
https://github.com/filecoin-project/lotus/issues/10414

## Proposed Changes
1. use StateCompute instead of StateSearchMsg for feeHistory
2. return MinGasPremium when there's no tx in the block

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -696,28 +696,30 @@ func (a *EthModule) EthFeeHistory(ctx context.Context, p jsonrpc.RawParams) (eth
 	rewardsArray := make([][]ethtypes.EthBigInt, 0)
 
 	for ts.Height() >= abi.ChainEpoch(oldestBlkHeight) {
-		// Unfortunately we need to rebuild the full message view so we can
-		// totalize gas used in the tipset.
-		msgs, err := a.Chain.MessagesForTipset(ctx, ts)
+		compOutput, err := a.StateCompute(ctx, ts.Height(), nil, ts.Key())
```
Comment (raulk) [inline node/impl/full/eth.go:699]: How does this deal with null rounds?

**Review by magik6k — APPROVED** [review]
Looks like this works. Also should be faster, I think.

**Review by raulk — APPROVED** [review]

File: itests/eth_fee_history_test.go [hunk 2]
```diff
@@ -22,13 +24,16 @@ func TestEthFeeHistory(t *testing.T) {
...
 	blockTime := 100 * time.Millisecond
 	client, _, ens := kit.EnsembleMinimal(t, kit.MockProofs(), kit.ThroughRPC())
-	ens.InterconnectAll().BeginMining(blockTime)
+	miner := ens.InterconnectAll().BeginMining(blockTime)
 
 	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
 	defer cancel()
 
 	// Wait for the network to create 20 blocks
 	client.WaitTillChain(ctx, kit.HeightAtLeast(20))
```
Comment (arajasek) [inline itests/eth_fee_history_test.go:33]: What happens if some of these rounds are null blocks? I can't tell whether the test assertions are still correct.

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -695,29 +691,31 @@ func (a *EthModule) EthFeeHistory(ctx context.Context, p jsonrpc.RawParams) (eth
 	gasUsedRatioArray := []float64{}
 	rewardsArray := make([][]ethtypes.EthBigInt, 0)
 
-	for ts.Height() >= abi.ChainEpoch(oldestBlkHeight) {
-		// Unfortunately we need to rebuild the full message view so we can
-		// totalize gas used in the tipset.
-		msgs, err := a.Chain.MessagesForTipset(ctx, ts)
+	for len(gasUsedRatioArray) < int(params.BlkCount) && ts.Height() > 0 {
+		compOutput, err := a.StateCompute(ctx, ts.Height(), nil, ts.Key())
```
[the remark this answers is not in the document]
Comment (raulk) [reply node/impl/full/eth.go:695]: I'm doing a sanity check on the logic here, it's ultimately in `parseBlkParam`.

[the remark this answers is not in the document]
Comment (raulk) [reply node/impl/full/eth.go:695]: https://github.com/filecoin-project/lotus/issues/10426 -- will be fixed separately

File: itests/eth_fee_history_test.go [hunk 4]
```diff
@@ -29,95 +46,109 @@ func TestEthFeeHistory(t *testing.T) {
...
+			for _, c := range chg {
+				tsHeights = append(tsHeights, int(c.Val.Height()))
+			}
+		}
+	}()
+
+	miner[0].InjectNulls(abi.ChainEpoch(1))
+	<-time.After(5 * blockTime)
+	miner[0].InjectNulls(abi.ChainEpoch(3))
+	<-time.After(2 * blockTime)
```
Comment (raulk) [inline itests/eth_fee_history_test.go:65]: Depending on time is usually flaky. Consider waiting for N number of tipsets produced again. See https://github.com/filecoin-project/lotus/pull/10427/files#diff-57c54de2d685e694dee103dc15a360776a5b30b5ec7fc77b5150c3756525d683R322-R328 for some hints.

File: itests/eth_fee_history_test.go [hunk 5]
```diff
@@ -29,95 +46,109 @@ func TestEthFeeHistory(t *testing.T) {
...
+	miner[0].InjectNulls(abi.ChainEpoch(1))
+
+	// Wait for the network to create at least 20 blocks
```
Comment (raulk) [inline itests/eth_fee_history_test.go:68]: Tipsets, not blocks.

File: itests/eth_fee_history_test.go [hunk 6]
```diff
@@ -29,95 +46,109 @@ func TestEthFeeHistory(t *testing.T) {
...
 	client.WaitTillChain(ctx, kit.HeightAtLeast(20))
 	for _, m := range miner {
 		m.Pause()
 	}
+	<-time.After(5 * blockTime)
```
Comment (raulk) [inline itests/eth_fee_history_test.go:73]: This is racy. It's possible that the miners may have not pausd yet. Safer to do a ChainNotify with a 5 second timeout and assert that it errors with context deadline exceeded.

File: itests/eth_fee_history_test.go [hunk 7]
```diff
@@ -29,95 +46,109 @@ func TestEthFeeHistory(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
 	defer cancel()
 
-	// Wait for the network to create 20 blocks
+	heads, err := client.ChainNotify(ctx)
+	require.NoError(err)
+
+	// Save the full view of the tipsets to calculate the answer when there are null rounds
+	tsHeights := []int{1}
```
Comment (raulk) [inline itests/eth_fee_history_test.go:53]: Why does this need to be preloaded with 1?

File: itests/eth_fee_history_test.go [hunk 8]
```diff
@@ -29,95 +46,109 @@ func TestEthFeeHistory(t *testing.T) {
...
+	go func() {
+		for chg := range heads {
+			for _, c := range chg {
+				tsHeights = append(tsHeights, int(c.Val.Height()))
```
Comment (raulk) [inline itests/eth_fee_history_test.go:57]: Note that the first notification you receive will be the current tipset, i.e. c.Type will be store.HCCurrent.

File: itests/eth_fee_history_test.go [hunk 9]
```diff
@@ -29,95 +46,109 @@ func TestEthFeeHistory(t *testing.T) {
...
+	// Wait for the network to create at least 20 blocks
 	client.WaitTillChain(ctx, kit.HeightAtLeast(20))
 	for _, m := range miner {
 		m.Pause()
 	}
+	<-time.After(5 * blockTime)
+
+	sort.Ints(tsHeights)
+
+	latestBlk := ethtypes.EthUint64(tsHeights[len(tsHeights)-2])
```
Comment (raulk) [inline itests/eth_fee_history_test.go:77]: Can you add a comment indicating that this is the last executed tipset?

File: itests/eth_fee_history_test.go [hunk 10]
```diff
@@ -17,6 +19,21 @@ import (
 	"github.com/filecoin-project/lotus/node/impl/full"
 )
 
+func getAnswer(tsHeights []int, requestAmount, startHeight int) (count, oldestHeight int) {
```
Comment (raulk) [inline itests/eth_fee_history_test.go:22]: This needs docs and comments. Can we rename to `calculateExpectations`?

File: node/impl/full/eth.go [hunk 11]
```diff
@@ -252,7 +252,7 @@ func (a *EthModule) parseBlkParam(ctx context.Context, blkParam string) (tipset
 		if err != nil {
 			return nil, fmt.Errorf("cannot parse block number: %v", err)
 		}
-		ts, err := a.Chain.GetTipsetByHeight(ctx, abi.ChainEpoch(num), nil, false)
+		ts, err := a.Chain.GetTipsetByHeight(ctx, abi.ChainEpoch(num), nil, true)
```
Comment (raulk) [inline node/impl/full/eth.go:255]: Can you revert this change, since it's already in https://github.com/filecoin-project/lotus/pull/10427?

File: node/impl/full/eth.go [hunk 12]
```diff
@@ -691,7 +691,8 @@ func (a *EthModule) EthFeeHistory(ctx context.Context, p jsonrpc.RawParams) (eth
 	gasUsedRatioArray := []float64{}
 	rewardsArray := make([][]ethtypes.EthBigInt, 0)
 
-	for len(gasUsedRatioArray) < int(params.BlkCount) && ts.Height() > 0 {
+	blockCount := 0
```
Comment (raulk) [inline node/impl/full/eth.go:694]: Let's call this `blocksIncluded`.

File: itests/eth_fee_history_test.go [hunk 13]
```diff
@@ -17,6 +19,21 @@ import (
...
+	idx := sort.SearchInts(tsHeights, startHeight)
```
Comment (raulk) [inline itests/eth_fee_history_test.go:23]: latestIdx

File: itests/eth_fee_history_test.go [hunk 14]
```diff
@@ -17,6 +19,21 @@ import (
...
+	// handle null rounds
+	for tsHeights[idx] > startHeight {
+		idx--
+	}
+	cnt := requestAmount
+	oldestIdx := idx - requestAmount + 1
+	if idx < cnt {
+		cnt = idx + 1
```
Comment (raulk) [inline itests/eth_fee_history_test.go:31]: Not sure I follow this.

If you have `tsHeights = [1, 2, 3, 4, 6, 8]`, and you request `startHeight = 5`, you will have `idx = 4` (current value 6). If you request 10 heights, you will enter this if, and will end up returning cnt = 5, even though you would only process 4 heights (1..4)?

File: itests/eth_fee_history_test.go (hunk 7 above)
Comment (ychiaoli18) [reply itests/eth_fee_history_test.go:53]: It seems that ChainNotify doesn't send the tipset with height = 1

File: itests/eth_fee_history_test.go (hunk 14 above)
Comment (ychiaoli18) [reply itests/eth_fee_history_test.go:31]: In your example, idx would be 3 (idx would be the index of the largest number that's smaller than the target), so it would return 4 entries. I can rewrite it to make it clearer.

File: node/impl/full/eth.go (hunk 11 above)
Comment (ychiaoli18) [reply node/impl/full/eth.go:255]: The e2e test would fail if this line is not included. I can rebase if #10427 is merged.

File: itests/eth_fee_history_test.go (hunk 14 above)
Comment (raulk) [reply itests/eth_fee_history_test.go:31]: Hm, not according to the Godocs and this playground: https://play.golang.com/p/bQyjTqj9Gxb

Comment (ychiaoli18) [reply itests/eth_fee_history_test.go:31]: There's a [for loop](https://github.com/filecoin-project/lotus/pull/10413/files/1747893ead537a24b39082c2f7000ceb742b5e13..61d2d461e1e4dc29b360f095d070b48377922c7d#diff-4ec74518af0f29b47821833fc7b7ff99d23cb2b21ce3bfe7deeb6c9994f64646R25-R27) that decreases the idx

File: itests/eth_fee_history_test.go (hunk 6 above)
Comment (raulk) [reply itests/eth_fee_history_test.go:73]: Hm, that won't work, but we should wait for 5 seconds of ChainNotify not sending any notifications.

Comment (raulk) [reply itests/eth_fee_history_test.go:73]: Honestly I think this is fine, just keep it as it is. Pause is synchronous, so we're just waiting on draining all pending chain notifications, and 5 seconds is of plenty time for that.

File: itests/eth_fee_history_test.go (hunk 7 above)
Comment (raulk) [reply itests/eth_fee_history_test.go:57]: Hm, that suggests raciness. The ChainNotify is probably starting on tipset 2. Note that ChainNotify sends a "current" notification on subscribe, and the subscription might be starting on height 2. Do we absolutely need to use ChainNotify here? Alternatively you can walk the chain and populate tsHeights after the fact, i.e. after pausing the miner?

File: node/impl/full/eth.go (hunk 11 above)
Comment (raulk) [reply node/impl/full/eth.go:255]: Gotcha, thanks.

File: itests/eth_fee_history_test.go (hunk 7 above)
Comment (raulk) [reply itests/eth_fee_history_test.go:57]: ChainNotify being called before the miner starts resolves my raciness concerns. Starting at 2 is probably an artifact from the testkit, and the way we set up miners.

**Review by raulk — APPROVED** [review]
